corehq/apps/sso/tests/test_tasks.py

```diff
@@ -133,3 +139,103 @@ def tearDown(self):
+            cls.account,
+            cls.domain,
+            date_start=datetime.date.today(),
+            date_end=None,
+            plan_version=enterprise_plan,
+            is_active=True,
+        )
+
+        cls.idp = generator.create_idp('vaultwax', cls.account)
+        cls.idp.enable_user_deactivation = True
+        cls.idp.idp_type = IdentityProviderType.AZURE_AD
+        cls.idp.save()
+        cls.email_domain = AuthenticatedEmailDomain.objects.create(
+            email_domain='vaultwax.com',
+            identity_provider=cls.idp,
+        )
+
+    def setUp(self):
+        super().setUp()
+        self.web_user_a = self._create_web_user('a@vaultwax.com')
+        self.web_user_b = self._create_web_user('b@vaultwax.com')
+        self.web_user_c = self._create_web_user('c@vaultwax.com')
+        self.web_user_d = self._create_web_user('d@vaultwax.com')
+
+    @patch('corehq.apps.sso.models.IdentityProvider.get_all_members_of_the_idp')
+    def test_user_deactivation_logic(self, mock_get_all_members_of_the_idp):
+        self.assertTrue(self.web_user_c.is_active)
+        mock_get_all_members_of_the_idp.return_value = [self.web_user_a.username, self.web_user_b.username]
+
+        auto_deactivate_removed_sso_users()
+
+        # Refetch Web User
+        web_user = WebUser.get_by_username(self.web_user_c.username)
+        self.assertFalse(web_user.is_active)
+
+    @patch('corehq.apps.sso.models.IdentityProvider.get_all_members_of_the_idp')
+    def test_sso_exempt_users_are_not_deactivated(self, mock_get_all_members_of_the_idp):
+        sso_exempt = self._create_web_user('exempt@vaultwax.com')
+        UserExemptFromSingleSignOn.objects.create(
+            username=sso_exempt.username,
+            email_domain=self.email_domain,
+        )
+        mock_get_all_members_of_the_idp.return_value = [self.web_user_a.username, self.web_user_b.username]
+
+        auto_deactivate_removed_sso_users()
+
+        # Refetch Web User
+        web_user = WebUser.get_by_username(sso_exempt.username)
+        self.assertTrue(web_user.is_active)
+
+    @patch('corehq.apps.sso.models.IdentityProvider.get_all_members_of_the_idp')
+    @patch('corehq.apps.sso.tasks.send_html_email_async.delay')
+    def test_deactivation_skipped_if_azure_return_empty_sso_user(self, mock_send, mock_get_all_members_of_the_idp):
+        mock_get_all_members_of_the_idp.return_value = []
+
+        auto_deactivate_removed_sso_users()
+
+        # Refetch Web User
+        web_user_a = WebUser.get_by_username(self.web_user_a.username)
+        self.assertTrue(web_user_a.is_active)
+        web_user_b = WebUser.get_by_username(self.web_user_b.username)
+        self.assertTrue(web_user_b.is_active)
+        web_user_c = WebUser.get_by_username(self.web_user_c.username)
+        self.assertTrue(web_user_c.is_active)
+        mock_send.assert_called_once()
+
+    @patch('corehq.apps.sso.models.IdentityProvider.get_all_members_of_the_idp')
```

Comment:
super nit: Since you are patching this for every test, setting this patcher up at the class level could help make each test a bit more concise and readable. That could just be moving the decorator to the class, but I like setting up the patcher in setUp (or setUpClass), because you avoid having the extra parameter in each test method. Something like this:
```
        idp_patcher = patch('corehq.apps.sso.models.IdentityProvider.get_all_members_of_the_idp')
        cls.mock_get_members_of_idp = idp_patcher.start()
        cls.addClassCleanup(idp_patcher.stop)
```

Reply:
Thank you! Love this suggestion. Refactored here: 080b22a5c40adab3718728ed3261e3ad3afccfdb